src/RecordSession.cc

```diff
@@ -120,13 +120,23 @@ static void record_exit_trace_event(RecordTask* t, WaitStatus exit_status) {
  * true, t has been deleted and cannot be referenced again.
  */
 static bool handle_ptrace_exit_event(RecordTask* t) {
+  if (t->already_reaped()) {
+    if (!t->waiting_for_reap)
+      delete t;
+    return true;
```

Comment:
What's happening in this state? `already_reaped()` is true, but `waiting_for_reap` is also true?

Reply:
`already_reaped` refers to the actual state of the child, while `waiting_for_reap` refers to the state that we're presenting to the the emulated ptracer. That said, now that I think about it, we may want to avoid actually reaping or detaching from the child until the emulated ptracer has itself reaped the child, since reaping the child from the ptracer will make the exit visible to non-ptrace parent. That's not a behavior change in this commit though. We always detached the task right away. My preference would be to fix that in another commit with a proper test for the behavior.